src/psyclone/psyir/symbols/datatypes.py

```diff
@@ -909,7 +910,7 @@ class StructureType(DataType):
 
     '''
     @dataclass(frozen=True)
-    class ComponentType:
+    class ComponentType(CommentableMixin):
```

Comment:
What I would like to try is keeping this inheritance line but reverting back everything else modified inside this class. Then when you use it, e.g. in fparser, you will need to use:
```
dtype.add(symbol.name, datatype, symbol.visibility,
                  initial_value)
symbol.preceding_comment
symbol.inline_comment
```
and also you may need a to redefine `__copy__()` to do the same. But this is the same approach that is used for all the symbols `copy` method.

Would that work?

Reply:
Sadly this does not work because `preceding_comment` and `inline_comment` are immutable, due to the dataclass being frozen:
```
src/psyclone/psyir/symbols/datatypes.py:1055: in add
    self._components[name].preceding_comment = preceding_comment

E   dataclasses.FrozenInstanceError: cannot assign to field 'preceding_comment'
```

I get the same error when trying to redefine the setters using the ugly `object.__setattr__` hack.

The best I managed to do is using twice in `StructureType.add` instead of redefining `ComponentType.__init__`.

Reply:
Let me know what you think about this one :)

Reply:
Ok, I am happy with this solution